### DISCUSSION TITLE
Add primary key to prisma generated mapping tables

### DISCUSSION THREAD

**Discussion by webnoob** [prisma/prisma#11028] [Q&A] [answered]
Hi,

I'm using a managed DB hosting provider (OVH) and in turn have no control over the `sql_require_primary_key` setting. When importing the prisma generated DB, I get an error saying each table must have a primary key. For the most part, this is fine but I'm struggling with the tables that prisma creates as mapping tables.

Take this model:

```
model User {
  id          Int      @id @default(autoincrement())
  canActAs    User[] @relation("actingAs", references: [id])
  hasActingAs User[] @relation("actingAs", references: [id])
}
```

Prisma will create a `_actingas` table with 2 columns, `A` and `B`. I need to be able to add a PK field to that table.

How can I achieve this?

Note: I tried creating an empty migration which altered those mapping tables adding a PK field but it created a new migration removing it after because the schema was out of line.

Note: Initially, I didn't think those tables were causing my error because when I  look at the `_actingas` mapping table, I do see that both columns `A` and `B` are down as PK fields:
![image](https://user-images.githubusercontent.com/10404747/148643815-333df5c0-3f88-4789-b926-7ec00cb567c5.png)
 but the following query raises them as not having a PK field:

```
select tab.table_schema as database_name,
       tab.table_name
from information_schema.tables tab
left join information_schema.table_constraints tco
          on tab.table_schema = tco.table_schema
          and tab.table_name = tco.table_name
          and tco.constraint_type = 'PRIMARY KEY'
where tco.constraint_type is null
      and tab.table_schema not in('mysql', 'information_schema', 
                                  'performance_schema', 'sys')
      and tab.table_type = 'BASE TABLE'
--      and tab.table_schema = 'sakila' -- put schema name here
order by tab.table_schema,
         tab.table_name;
```

**Comment by austincrim** [comment, marked as answer]
Hey @webnoob 👋 ,

The only solution for now would be to switch to an **explicit** many-to-many relationship with a join table defined in your schema.

We are actively working on improving our m-to-n relations, so would you mind creating a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with the relevant details? Our engineering teams will be able to pick it up from there!

**Reply by webnoob** [reply]
Thanks for the response. I thought a manual approach might be the only way.
Thankfully OVH updated the server setting for me so it's no longer an issue, however, this point still stands.

Created an issue: https://github.com/prisma/prisma/issues/11110 (hope the format is Ok, I used the create from discussion option and updated the wording)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
